### ISSUE TITLE
Use `touch EXIT` to stop CP2K at walltime end

### ISSUE THREAD

**Issue by danieleongari** [aiidateam/aiida-cp2k#51]
CP2K allows for a soft kill, by creating in the running folder an empty file called `EXIT`.
This solution con be implemented in the plugin to stop the code. 

When using `EXIT`, CP2K stops at the current SCF step (returning a warning as if it did not converge) but printing the standard termination of the program, and therefore allowing for a smoother parsing of the output.

**Comment by yakutovicha** [comment]
Another possibility here would be to specify the `walltime` in the cp2k input slightly smaller then it is set in the batch file.

**Comment by yakutovicha** [comment]
I am not sure, though, if the input plugin can do something while the calculation is running. I am actually pretty sure it **can't**. @ltalirz do you think it is possible?

**Comment by yakutovicha** [comment]
Made an issue upstream: https://github.com/aiidateam/aiida-core/issues/3868. Let's see how this evolves.

**Comment by sphuber** [comment]
> Another possibility would be to specify the `walltime` parameter in the cp2k input slightly smaller then it is set in the batch file.

If CP2K provides this functionality, I would definitely go for this route. This is what I do in Quantum ESPRESSO as well and works relatively well. I implement this directly on the `PwBaseWorkChain`, where I always take the `metadata.options.resources.walltime_seconds` input and set a fraction of that in the input parameters of the `PwCalculation`. This way, all other workflow always automatically inherit this behavior.

I responded to the issue Sasha opened and although the other possibility is in principle possible to be implemented, it is quite challenging I would say.

**Comment by dev-zero** [comment]
Would this also mean that if I'm going to kill a process the base workchain could possibly react to it by first trying to write an `EXIT` file to terminate the process on remote gracefully?

**Comment by sphuber** [comment]
> Would this also mean that if I'm going to kill a process the base workchain could possibly react to it by first trying to write an `EXIT` file to terminate the process on remote gracefully?

Not really. The workchain execution is blocked until the child process (the `CalcJob` in this example) is terminated. That means the workchain cannot retake control and perform an action before the `CalcJob` has finished running. You could of course manually write an `EXIT` file in the working directory of the `CalcJob` which would cause the code to stop gracefully. The daemon will then realize the job is done and start retrieval and parsing. If the parser properly recognizes the graceful shutdown and sets an appropriate exit code, and the base restart has a handler for that exit code to simply start a new calcjob, restarting from the output of the last, then that works. That is exactly what the `PwBaseWorkChain` and `PwCalculation` do.
